File: Languages/npeg_php/AstNode.php
<?
namespace RobustHaven\Text\Npeg;

class AstNode 
{
	private $parent;
	private $children;
	private $token;

	function __construct() 
	{
       $this->children = array();
    }
    
	public function setParent( $value )
	{
		$this->parent = $value;
	}

	public function getParent()
	{
		return $this->parent;
	}

	public function setChildren( $value )
	{
		$this->children = $value;
	}

	public function getChildren()
	{
		return $this->children;
	}
	
	public function setToken( $value )
	{
		$this->token = $value;
	}

	public function getToken()
	{
		return $this->token;
	}
	
	public function Accept( $visitor)
    {
    	$visitor->VisitEnter($this);
		$isFirstTime = true;
        foreach ($this->Children as $node)
        {
        	if (!$isFirstTime)
            {
            	$visitor->VisitExecute($this);
            }
            $isFirstTime = false;
            $node->Accept($visitor);
        }
        $visitor->VisitLeave($this);
	}
}
?>

File: Languages/npeg_csharp/RobustHaven.Text.Npeg/AstNode.cs
using System;
using System.Collections.Generic;

namespace RobustHaven.Text.Npeg
{
	public class AstNode
	{
		public AstNode()
		{
			Children = new List<AstNode>();
		}

		public AstNode Parent { get; set; }

		public List<AstNode> Children { get; set; }

		public TokenMatch Token { get; set; }


		public void Accept(IAstNodeVisitor visitor)
		{
			visitor.VisitEnter(this);

			Boolean isFirstTime = true;
			foreach (AstNode node in Children)
			{
				if (!isFirstTime)
				{
					visitor.VisitExecute(this);
				}

				isFirstTime = false;
				node.Accept(visitor);
			}

			visitor.VisitLeave(this);
		}
	}
}